Add SI_STATUS_DMA_BUSY

loganmc10 opened the pull request:
This fixes the error:
```
Core Warning: two events of type 0x8 in interrupt queue
```
in some games. I've noticed it in Diddy Kong Racing and Mischief Makers (it really spams this message in Mischief Makers).

Anyway they are doing this because they don't know the SI DMA is busy. I added the busy bit, now these errors don't show up anymore.

loganmc10 commented:
I also added SI_STATUS_RD_BUSY, which is only applied for ```dma_pi_write()``` (a write to SI_PIF_ADDR_WR64B_REG). Perhaps someone could confirm that is correct before this is merged.

rcp.h says that SI_STATUS_RD_BUSY means "IO read busy". It's a little clear to me what that means, but I assume is means "when dram is being read from", which would be the case for ```dma_pi_write()```

loganmc10 commented:
Actually now I'm not sure, wouldn't the "IO" be the data from the PIF? In that case the bit should be set during ```dma_si_read``` @bsmiles32 do you know how this is supposed to work?

bsmiles32 commented:
I don't really know how SI_STATUS_RD_BUSY should be interpreted.
While these changes fix the "two event of type 0x8 in interrupt queue" spamming, it breaks button presses on my end in Mischief Maker.

loganmc10 commented:
Hm ok thanks for testing, I'll do some more testing and research

loganmc10 commented:
@bsmiles32 do you know why 0x900 was chosen as the interrupt delay time? Or what games need/benefit from the SI delay?

Reducing the 0x900 number fixes Mischief Makers. In my reading, the peak DMA transfer rate for the RSP is 8 bytes per cycle. So I have a couple questions:

What is the 0x900 number supposed to represent? What is it counting?

If it is supposed to be counting ops or cycles, is ```PIF_RAM_SIZE / 8``` a good approximation? It is a very small number (0x8), I can test this on a few games, I'm just wondering if you had any insight into how the current system came to be.

loganmc10 commented:
Even setting it to 0x100 makes it work, which is the number in the comment, but I'm not sure why it's commented

loganmc10 commented:
Shortening the interrupt timing broke Body Harvest, so I'm assuming that's not a safe thing to do.

I've modified the PR to get rid of RD_BUSY, since I don't see any benefit of using it. I also added DelaySI as a INI param and disabled it for Mischief Makers. This fixes the input. Does this seem ok? Diddy Kong Racing seems to perform as normal. I'm not aware of any other game that had those error messages.

bsmiles32 commented:
Don't know about the 0x900 value nor the delay si behavior. I guess it's pure try and guess what worked for selected titles (like for any timing stuff in m64p).
I'm not sure what speed the SI DMA has, so a test ROM is what we need here.

The updated PR looks OK to me, but some more testing is welcome before I merge it.

loganmc10 commented:
I guess this was tried once before:
https://github.com/mupen64plus/mupen64plus-core/commit/252d09f30cfd53a5be6f069e124e6051d03041b4

And reverted for this same reason:
https://github.com/mupen64plus/mupen64plus-core/commit/b77d5512db99650ab435f05daae02c629081c542

I'll do a fair bit of testing for this change to make sure it doesn't cause issues in any other games

loganmc10 commented:
I tested this with about 60 games. I made sure the input worked in all of them. I also made sure that input and saving works in Diddy Kong Racing and Mischief Makers.

It all looks good to me, no issues. If we ever run across another game like Mischief Makers that has issues, disabling DelaySI will probably do the trick, and now it can just be added to the ini file.
